[Windows] When adding font_features to settings, zed crashes until settings removed

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

## Description
Adding font features (`ui_font_features` | `buffer_font_features`) causes a crash.

tested properties:
- `italic` and `underlined` crashes
- example on the [zed docs](https://zed.dev/docs/configuring-zed#buffer-font-features) does not

## Reproduce
- open settings
- add e.g. `"buffer_font_features": {"italic":true}`
- zed crashes

## Error message when using "cargo run" with RUST_BACKTRACE
[error_msg.txt](https://github.com/user-attachments/files/15974543/error_msg.txt)



### Environment

OS: Windows 11
Processor: Intel(R) Core(TM) i7-14700HX   2.10 GHz
RAM: 32GB

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>expand for logs...</summary>
//When starting zed after crash
2024-06-25T18:15:46+02:00 [INFO] ========== starting zed ==========
2024-06-25T18:15:46+02:00 [INFO] Use Segoe UI as UI font.
2024-06-25T18:15:46+02:00 [INFO] Using direct write text system.
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Using git binary path: None
2024-06-25T18:15:46+02:00 [ERROR] crates\zed\src/main.rs:362: SHELL environment variable is not assigned so we can't source login environment variables

Caused by:
    environment variable not found
2024-06-25T18:15:46+02:00 [INFO] extensions updated. loading 6, reloading 0, unloading 0
2024-06-25T18:15:46+02:00 [ERROR] crates\zed\src\zed.rs:716: Expected array of length 2
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] set status on client 0: Authenticating
2024-06-25T18:15:46+02:00 [INFO] set status on client 177145: Connecting
2024-06-25T18:15:46+02:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-06-25T18:15:46+02:00 [INFO] Enabling Vulkan Portability
2024-06-25T18:15:46+02:00 [INFO] Enabling color space support
2024-06-25T18:15:46+02:00 [INFO] Adapter "NVIDIA GeForce RTX 4060 Laptop GPU"
2024-06-25T18:15:46+02:00 [INFO] Ray tracing is supported
2024-06-25T18:15:46+02:00 [WARN] Requested size 1x1 is outside of surface capabilities
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-06-25T18:15:46+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-06-25T18:15:46+02:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_impl")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_transformed")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("blend_color")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_impl")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_transformed")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("blend_color")
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [ERROR] crates\fs\src/fs.rs:537: Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Using Windows API for pseudoconsole
2024-06-25T18:15:47+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:47+02:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
</details>

--- a/crates/gpui/src/text_system/font_features.rs
+++ b/crates/gpui/src/text_system/font_features.rs
@@ -144,15 +144,14 @@ impl schemars::JsonSchema for FontFeatures {
 }
 
 fn is_valid_feature_tag(tag: &str) -> bool {
-    let raw_bytes = tag.bytes();
-    if raw_bytes.len() != 4 {
+    if tag.len() != 4 {
         return false;
     }
-    for ch in raw_bytes.into_iter() {
+    for ch in tag.chars() {
         match ch {
-            48..=57 => continue,  // 0 ... 9
-            65..=90 => continue,  // A ... Z
-            97..=122 => continue, // a ... z
+            '0'..='9' => continue,
+            'A'..='Z' => continue,
+            'a'..='z' => continue,
             _ => return false,
         }
     }

--- a/crates/gpui/src/text_system/font_features.rs
+++ b/crates/gpui/src/text_system/font_features.rs
@@ -58,7 +58,7 @@ impl<'de> serde::Deserialize<'de> for FontFeatures {
                 while let Some((key, value)) =
                     access.next_entry::<String, Option<FeatureValue>>()?
                 {
-                    if key.len() != 4 || !key.is_ascii() {
+                    if !is_valid_feature_tag(&key) {
                         log::error!("Incorrect font feature tag: {}", key);
                         continue;
                     }
@@ -141,4 +141,20 @@ impl schemars::JsonSchema for FontFeatures {
         }
         schema.into()
     }
+}
+
+fn is_valid_feature_tag(tag: &str) -> bool {
+    let raw_bytes = tag.bytes();
+    if raw_bytes.len() != 4 {
+        return false;
+    }
+    for ch in raw_bytes.into_iter() {
+        match ch {
+            48..=57 => continue,  // 0 ... 9
+            65..=90 => continue,  // A ... Z
+            97..=122 => continue, // a ... z
+            _ => return false,
+        }
+    }
+    true
 }

--- a/crates/gpui/src/text_system/font_features.rs
+++ b/crates/gpui/src/text_system/font_features.rs
@@ -148,11 +148,8 @@ fn is_valid_feature_tag(tag: &str) -> bool {
         return false;
     }
     for ch in tag.chars() {
-        match ch {
-            '0'..='9' => continue,
-            'A'..='Z' => continue,
-            'a'..='z' => continue,
-            _ => return false,
+        if !ch.is_ascii_alphanumeric() {
+            return false;
         }
     }
     true

--- a/crates/gpui/src/text_system/font_features.rs
+++ b/crates/gpui/src/text_system/font_features.rs
@@ -58,7 +58,7 @@ impl<'de> serde::Deserialize<'de> for FontFeatures {
                 while let Some((key, value)) =
                     access.next_entry::<String, Option<FeatureValue>>()?
                 {
-                    if key.len() != 4 && !key.is_ascii() {
+                    if key.len() != 4 || !key.is_ascii() {
                         log::error!("Incorrect font feature tag: {}", key);
                         continue;
                     }

